storage/src/main/scala/coop/rchain/storage/regex/Fsm.scala

```diff
@@ -0,0 +1,811 @@
+
+  /**
+    * True if FSM alphabet contains Fsm.AnythingElse and thus can accept unknown characters
+    */
+  val HasAnythingElse: Boolean = alphabet.contains(Fsm.AnythingElse)
+
+  /**
+    * A state is "live" if a final state can be reached from it.
+    */
+  def IsLive(state: Int): Boolean = {
+    val reachable = Queue(state)
+    val checked = HashSet(state)
+    //can't use any functional iterator here (like reachable.exists), since queue is modified
+    while (reachable.nonEmpty) {
+      val reachableState = reachable.dequeue()
+
+      if (finalStates.contains(reachableState)) {
+        //we got the result, no more iterations needed, return immediately
+        return true
+      }
+
+      checked += reachableState
+
+      transitions.get(reachableState) match {
+        case Some(charToStateMap) =>
+          charToStateMap.foreach(charToState => {
+            val (_, transitionTargetState) = charToState
+            if (!checked.contains(transitionTargetState)) {
+              reachable += transitionTargetState
+            }
+          })
+        case None =>
+      }
+    }
+
+    false
+  }
+
+  /**
+    * Test whether the present FSM accepts the supplied string (iterable of symbols).
+    * Equivalently, consider `self` as a possibly-infinite set of
+    * strings and test whether `string` is a member of it.
+    * This is actually mainly used for unit testing purposes.
+    * If `Fsm.AnythingElse` is in your alphabet, then any symbol not in your
+    * alphabet will be converted to `Fsm.AnythingElse`
+    */
+  def Accepts(input: String): Boolean = {
+    var currentState = initialState
+    for (currentSymbol <- input) {
+      val sym =
+        if (HasAnythingElse & !alphabet.contains(currentSymbol))
+          Fsm.AnythingElse
+        else currentSymbol
+
+      transitions.get(currentState) match {
+        case None =>
+          //Current state doesn't exist, FSM is Null or similar
+          return false
+        case Some(charToStateMap) => {
+          charToStateMap.get(sym) match {
+            case Some(nextState) => currentState = nextState
+            case None =>
+              //Missing transition = transition to dead state
+              return false
+          }
+        }
+      }
+    }
+
+    finalStates.contains(currentState)
+  }
+
+  /**
+    * Compute the Brzozowski derivative of this FSM with respect to the input
+    * string of symbols. <https://en.wikipedia.org/wiki/Brzozowski_derivative>
+    * If any of the symbols are not members of the alphabet, that's a NoSuchElementException.
+    * If you fall into oblivion, then the derivative is an FSM accepting no
+    * strings.
+    */
+  def Derive(input: String): Fsm = {
+    var currentState = initialState
+    for (currentSymbol <- input) {
+      val sym = if (alphabet.contains(currentSymbol)) {
+        currentSymbol
+      } else if (HasAnythingElse) {
+        Fsm.AnythingElse
+      } else {
+        throw new NoSuchElementException(
```

Comment:
I think we don't want to use exceptions, but we could use `Either[Error,Result]` as function result type instead.

Reply:
After investigating usage patterns for Either, Option and Try, I decided to use Try[Fsm] here.